### ISSUE TITLE
Getting SAT when comparing a function to itself

### ISSUE THREAD

**Issue by fortunac** [draperlaboratory/cbat_tools#115]
I'm currently testing `wp` by comparing the function `version_etc_ar` to itself in `base64`. Because it is the same exact function, I am expecting a result of `UNSAT`, but we are getting `SAT`.

My current hypothesis is that there is a bug when we chaos the caller-saved registers at a function call with a predicate that takes in the input registers as arguments.

The command I'm running is:
```
  bap base64-gcc --pass=wp \
    --wp-compare=true \
    --wp-file1=base64-gcc.bpj \
    --wp-file2=base64-gcc.bpj \
    --wp-function=version_etc_ar \
    --wp-num-unroll=0 \
    --wp-output-vars=RAX,RCX,RDX,RSI,RDI,R8,R9,R10,R11 \
    --wp-use-fun-input-regs=true \
    --no-byteweight
```
And the refuted goal is:
```
R80 = R80152:
    Concrete values: = #x0000000000000001 #x0000000000000000
    Z3 Expression:
    (=
      (version_etc_arn_ret_R8
        RDI0
        RSI0
        RDX0
        RCX0
        R80
        #x0000000000000001)
      (version_etc_arn_ret_R8
        RDI0308
        RSI0309
        RDX0310
        RCX0311
        R80152
        #x0000000000000001))
```

**Comment by fortunac** [comment]
Upon closer inspection, we are getting `SAT` due to two things:
1. The only registers used in the subroutine `version_etc_ar` are `R8` and `R9`, and as a result when we generate the postcondition and hypotheses we only compare these two registers between the binaries. However, the function predicates include `RDI`, `RSI`, `RDX`, and `RCX`. Since they aren't included in the hypotheses, they don't have to be equal between both binaries. We should probably add them when we have input registers as function predicates.
2. Our current memory model says that memory is equal at an offset if the read address is on the heap, and that memory is equal if the address on the stack. In this case, the memory read is at the pointer stored in `R8` with `R9` as the index. Since `R8` is used as an argument to this subroutine, we have no idea whether that read is on the stack or heap, thus we are generating a countermodel where the memory is different at this address.

**Comment by fortunac** [comment]
Fixed with #120
